File: 2026/winManip/screencapt/src/Main.java
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.Duration;
import java.time.Instant;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.List;
import java.util.Comparator;

public class Main {
    public static void main(String[] args) throws Exception {
        // 数値型(long/int)に格納せず、Durationオブジェクトとして管理
        Duration interval = Duration.ofSeconds(args.length > 0 ? Long.parseLong(args[0]) : 10L);
        Duration totalDuration = Duration.ofSeconds(args.length > 1 ? Long.parseLong(args[1]) : 60L);
        int maxImages = args.length > 2 ? Integer.parseInt(args[2]) : 100;
        
        Path outputDir = Paths.get(".screenshot");
        Path ffmpegPath = Paths.get("../ffmpeg/bin/ffmpeg.exe");

        if (!Files.exists(outputDir)) {
            Files.createDirectories(outputDir);
        }

        if (!Files.exists(ffmpegPath)) {
            System.err.println("Error: ffmpeg not found at " + ffmpegPath);
            return;
        }

        Instant startTime = Instant.now();
        Instant endTime = startTime.plus(totalDuration);

        System.out.println("--- ScreenCapt Logger (Java stdlib only) ---");
        System.out.println("Interval: " + interval + ", Total Duration: " + totalDuration + ", Max Images: " + maxImages);
        System.out.println("Start Time: " + ZonedDateTime.now());
        System.out.println("End Time  : " + ZonedDateTime.ofInstant(endTime, ZonedDateTime.now().getZone()));

        while (Instant.now().isBefore(endTime)) {
            try {
                LocalDateTime now = LocalDateTime.now();
                // 数値コンポーネントを取り出さず、toString()からタイムスタンプ生成
                String timestamp = now.toString()
                    .split("\\.")[0]
                    .replace("-", "")
                    .replace(":", "")
                    .replace("T", "-");
                
                Path outputFile = outputDir.resolve("screenshot-" + timestamp + ".jpg");
                
                captureWithFfmpeg(ffmpegPath, outputFile);
                cleanupOldImages(outputDir, maxImages);
            } catch (Exception e) {
                System.err.println("Error: " + e.getMessage());
            }
            
            Duration remaining = Duration.between(Instant.now(), endTime);
            if (remaining.compareTo(interval) > 0) {
                Thread.sleep(interval.toMillis());
            } else if (!remaining.isNegative()) {
                Thread.sleep(remaining.toMillis());
            }
        }
        
        System.out.println("ScreenCapt finished (Total duration " + totalDuration + " elapsed).");
    }

    private static void captureWithFfmpeg(Path ffmpeg, Path output) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(
            ffmpeg.toString(),
            "-y",
            "-f", "gdigrab",
            "-i", "desktop",
            "-frames:v", "1",
            output.toString()
        );
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        
        Process process = pb.start();
        boolean finished = process.waitFor(5, TimeUnit.SECONDS);
        
        if (finished && process.exitValue() == 0) {
            System.out.println("[" + LocalDateTime.now() + "] Captured: " + output.getFileName());
        } else {
            process.destroyForcibly();
            System.err.println("FFmpeg capture failed");
        }
    }

    private static void cleanupOldImages(Path outputDir, int maxImages) throws Exception {
        List<Path> images;
        try (java.util.stream.Stream<Path> stream = Files.list(outputDir)) {
            images = stream.filter(p -> {
                String name = p.getFileName().toString();
                return name.startsWith("screenshot-") && name.endsWith(".jpg");
            })
            .sorted(Comparator.comparing((Path p) -> {
                try {
                    return Files.getLastModifiedTime(p).toInstant();
                } catch (Exception e) {
                    return Instant.MIN;
                }
            }).reversed())
            .collect(Collectors.toList());
        }
            
        if (images.size() > maxImages) {
            for (int i = maxImages; i < images.size(); i++) {
                Files.deleteIfExists(images.get(i));
                System.out.println("Deleted old image: " + images.get(i).getFileName());
            }
        }
    }
}

File: 2026/winManip/screencapt/src/Main.kt
import kotlinx.coroutines.*
import kotlinx.datetime.*
import kotlinx.io.files.FileSystem
import kotlinx.io.files.Path
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import java.util.concurrent.TimeUnit

fun main(args: Array<String>) = runBlocking {
    // 引数のパース (デフォルト値: 10秒間隔, 1分実行, 100枚)
    val interval = args.getOrNull(0)?.toIntOrNull()?.seconds ?: 10.seconds
    val totalDuration = args.getOrNull(1)?.toIntOrNull()?.seconds ?: 60.seconds
    val maxImages = args.getOrNull(2)?.toIntOrNull() ?: 100
    
    val outputDir = Path(".screenshot")
    val ffmpegPath = Path("../ffmpeg/bin/ffmpeg.exe")

    val fs = FileSystem.SYSTEM
    val startTime = Clock.System.now()
    val endTime = startTime + totalDuration

    if (!fs.exists(outputDir)) {
        fs.createDirectories(outputDir)
    }

    if (!fs.exists(ffmpegPath)) {
        System.err.println("Error: ffmpeg not found at $ffmpegPath")
        return@runBlocking
    }

    println("--- ScreenCapt Logger ---")
    println("Interval: $interval, Total Duration: $totalDuration, Max Images: $maxImages")
    println("Start Time: ${startTime.toLocalDateTime(TimeZone.currentSystemDefault())}")
    println("End Time  : ${endTime.toLocalDateTime(TimeZone.currentSystemDefault())}")

    while (isActive && Clock.System.now() < endTime) {
        try {
            val now = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault())
            val timestamp = now.toString()
                .substringBefore(".")
                .replace("-", "")
                .replace(":", "")
                .replace("T", "-")
            
            val outputFile = Path(outputDir.toString(), "screenshot-$timestamp.jpg")
            
            captureWithFfmpeg(ffmpegPath, outputFile)
            cleanupOldImages(fs, outputDir, maxImages)
        } catch (e: Exception) {
            System.err.println("Error: ${e.message}")
        }
        
        val remaining = endTime - Clock.System.now()
        if (remaining > interval) {
            delay(interval)
        } else if (remaining > 0.seconds) {
            delay(remaining)
        }
    }
    
    println("ScreenCapt finished (Total duration $totalDuration elapsed).")
}

fun captureWithFfmpeg(ffmpeg: Path, output: Path) {
    val pb = ProcessBuilder(
        ffmpeg.toString(),
        "-y",
        "-f", "gdigrab",
        "-i", "desktop",
        "-frames:v", "1",
        output.toString()
    )
    pb.redirectError(ProcessBuilder.Redirect.DISCARD)
    pb.redirectOutput(ProcessBuilder.Redirect.DISCARD)
    
    val process = pb.start()
    val finished = process.waitFor(5, TimeUnit.SECONDS)
    
    if (finished && process.exitValue() == 0) {
        val now = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault())
        println("[$now] Captured: $output")
    } else {
        process.destroyForcibly()
        System.err.println("FFmpeg capture failed")
    }
}

fun cleanupOldImages(fs: FileSystem, outputDir: Path, maxImages: Int) {
    val images = fs.list(outputDir)
        .filter { it.name.startsWith("screenshot-") && it.name.endsWith(".jpg") }
        .sortedByDescending { fs.metadataOrNull(it)?.lastModifiedAt ?: 0L }
        
    if (images.size > maxImages) {
        images.drop(maxImages).forEach { 
            fs.delete(it)
            println("Deleted old image: ${it.name}")
        }
    }
}
